src/main/kotlin/minesweeper/entity/MineField.kt

```diff
@@ -20,7 +20,52 @@ class MineField(
 
     fun countAroundMines(coordinate: Coordinate): Int {
         return coordinate.adjacentCoordinates()
-            .filter { it.isWithinBounds(width.value, height.value) }
+            .filter { it.isWithinBounds(width, height) }
             .count { _cells.findCell(it) is Cell.Mine }
     }
+
+    fun open(coordinate: Coordinate) {
+        val cell = findCell(coordinate)
+        if (cell.isRevealed) return
+
+        cell.open()
+        if (shouldOpenAdjacentCells(cell, coordinate)) {
+            openAdjacentEmptyCells(coordinate)
+        }
+    }
+
+    private fun shouldOpenAdjacentCells(
+        cell: Cell,
+        coordinate: Coordinate,
+    ): Boolean {
+        return cell is Cell.Empty && countAroundMines(coordinate) == 0
+    }
+
+    fun determineAction(coordinate: Coordinate): Action {
+        val cell = findCell(coordinate)
+
+        return when {
+            cell is Cell.Mine -> Action.GAME_OVER
+            isAllSafeCellsRevealed() -> Action.GAME_CLEARED
+            else -> Action.CONTINUE
+        }
+    }
+
+    private fun isAllSafeCellsRevealed(): Boolean {
+        return cells
+            .filterIsInstance<Cell.Empty>()
+            .all { it.isRevealed }
+    }
+
+    private fun openAdjacentEmptyCells(coordinate: Coordinate) {
+        coordinate.adjacentCoordinates()
+            .filter { it.isWithinBounds(width, height) && !findCell(it).isRevealed }
+            .forEach {
+                val cell = findCell(it)
+                cell.open()
+                if (cell is Cell.Empty && countAroundMines(cell.coordinate) == 0) {
```

Comment:
위에서 아래와 같은 코멘트를 남겼었는데요! 
> 생각해보니 어짜피 coordinate는 cell.coordinate와 동일할 것 같은데, shouldOpenAdjacentCells 메서드의 파라미터로 cell만 받아도 충분하지 않을까요?

이 코멘트대로 shouldOpenAdjacentCells 메서드의 파라미터가 수정된다면 이걸 재활용해볼 수 있을 것 같아요. 

```suggestion
                if (shouldOpenAdjacentCells(cell)) {
```